File: kotlin-node/karakum/src/jsMain/resources/plugins/ResourceRecordTypePlugin.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

function isRrtypeParameterType(node) {
    return (
        node.parent
        && ts.isLiteralTypeNode(node.parent)

        && node.parent.parent
        && ts.isParameter(node.parent.parent)
        && ts.isIdentifier(node.parent.parent.name)
        && node.parent.parent.name.text === "rrtype"

        && node.parent.parent.parent
        && (
            (
                ts.isFunctionDeclaration(node.parent.parent.parent)
                && node.parent.parent.parent.name
                && node.parent.parent.parent.name.text === "resolve"
            )
            || ts.isCallSignatureDeclaration(node.parent.parent.parent)
        )
    )
}

export default {
    setup(context) {
        this.sourceFileName = "generated.d.ts"
        this.namespace = undefined
        this.rrtypes = new Set()
    },

    traverse(node, context) {
        if (ts.isStringLiteral(node) && isRrtypeParameterType(node)) {
            const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())

            const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

            // generate the same entity for promise and callback API
            if (!sourceFileName.endsWith("promises.d.ts")) {
                this.sourceFileName = sourceFileName
                this.namespace = typeScriptService?.findClosest(node, ts.isModuleDeclaration)
            }

            this.rrtypes.add(node.text)
        }
    },

    render(node, context) {
        if (ts.isStringLiteral(node) && isRrtypeParameterType(node)) {
            return `ResourceRecordType.${karakum.constIdentifier(node.text)}`
        }

        return null
    },

    generate(context) {
        const name = "ResourceRecordType"

        const entries = Array.from(this.rrtypes).map(eventName => {
            const key = karakum.constIdentifier(eventName)
            return [key, eventName]
        })

        const keys = entries.map(([key]) => key)

        const body = keys
            .map(key => `sealed interface ${key} : ${name}`)
            .join("\n")

        const companionBody = entries
            .map(([key, value]) => (
                `
@seskar.js.JsValue("${value}")
val ${key}: ${key}
                `.trim()
            ))
            .join("\n")

        const fileBody = `
sealed external interface ${name} {
${body}

companion object {
${companionBody}
}
}
        `

        const declaration = {
            sourceFileName: this.sourceFileName,
            namespace: this.namespace,
            fileName: `${name}.kt`,
            body: fileBody,
        }

        return karakum.generateDerivedDeclarations([declaration], context);
    },
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/ResourceRecordTypePlugin.kt
package node.karakum.plugins

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.GeneratedFile
import io.github.sgrishchenko.karakum.extension.Plugin
import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.extension.ifPresent
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.structure.derived.DerivedDeclaration
import io.github.sgrishchenko.karakum.structure.derived.generateDerivedDeclarations
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import js.array.ReadonlyArray
import node.karakum.util.impure
import node.karakum.util.nullable
import typescript.*

private fun extractRrtypes(node: Node) = nullable {
    val parameter = ensureNotNull(node.getParentOrNull())
    ensure(isParameter(parameter))
    ensure(parameter.type === node)

    val parameterName = parameter.name
    ensure(isIdentifier(parameterName))
    ensure(parameterName.text == "rrtype")

    val resolveSignature = ensureNotNull(parameter.getParentOrNull())
    ensureNotNull(
        nullable {
            ensure(isFunctionDeclaration(resolveSignature))

            val functionName = ensureNotNull(resolveSignature.name)
            ensure(functionName.text == "resolve")
        } ?: nullable {
            ensure(isCallSignatureDeclaration(resolveSignature))
        }
    )

    nullable {
        ensure(isLiteralTypeNode(node))

        val literal = node.literal
        ensure(isStringLiteral(literal))

        listOf(literal.text)
    } ?: nullable {
        ensure(isUnionTypeNode(node))
        ensure(node.types.asArray().all { isLiteralTypeNode(it) })

        node.types.asArray().mapNotNull {
            nullable {
                ensure(isLiteralTypeNode(it))

                val literal = it.literal
                ensure(isStringLiteral(literal))

                literal.text
            }
        }
    }
}

class ResourceRecordTypePlugin : Plugin {
    private lateinit var sourceFileName: String
    private var namespace: ModuleDeclaration? = null
    private val rrtypes = mutableSetOf<String>()
    private val stringRrtypes = mutableSetOf<String>()
    private val promiseApiDeclarations = mutableListOf<DerivedDeclaration>()

    override fun setup(context: Context) = Unit

    override fun traverse(node: Node, context: Context) = impure {
        val types = ensureNotNull(extractRrtypes(node))

        rrtypes += types
        if (types.size > 1) stringRrtypes += types

        val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        // generate the same entity for promise and callback API
        ensure(!sourceFileName.endsWith("promises.d.ts"))

        this@ResourceRecordTypePlugin.sourceFileName = sourceFileName
        namespace = typeScriptService.findClosestNamespace(node)
    }

    override fun render(node: Node, context: Context, next: Render<Node>) = nullable {
        val types = ensureNotNull(extractRrtypes(node))

        nullable {
            ensure(types.size == 1)

            "ResourceRecordType.${types.first()}"
        } ?: run {
            "ResourceRecordType.StringRecord"
        }
    }

    override fun generate(context: Context, render: Render<Node>): ReadonlyArray<GeneratedFile> {
        val name = "ResourceRecordType"

        val body = rrtypes.joinToString("\n") {
            val marker = if (it in stringRrtypes) "StringRecord" else ""

            "sealed interface $it : $name${ifPresent(marker) { ", $marker"}}"
        }

        val companionBody = rrtypes.joinToString("\n") {
            """
                @seskar.js.JsValue("$it")
                val ${it}: $it
            """.trimIndent()
        }

        val fileBody = """
            sealed external interface $name {

            sealed interface StringRecord

            $body

            companion object {
            $companionBody
            }
            }
        """.trimIndent()

        val declaration = DerivedDeclaration(
            sourceFileName,
            namespace,
            fileName = "${name}.kt",
            fileBody,
        )

        return generateDerivedDeclarations(promiseApiDeclarations.toTypedArray() + declaration, context)
    }

}
